haskell/DemystifyingFoldable.hs

```diff
@@ -0,0 +1,45 @@
+{-
+
+Foldable represents a structure that, given a suitable "folding" function, can
+be collapsed into a new value without the structure.
+
+In these reps you'll practice the various ways Foldable can be implemented and
+the how to implement it on different kinds of datastructures.
+
+- Implement a type like Maybe
+- Write a full instance of Foldable for your "Maybe"
+- Write a newtype around List
+- Write a full instance of Foldable for your "List"
+
+-}
+
+
+module DemystifyingFoldable where
+
+data Option a
+  = Something a
+  | None
+
+instance Foldable Option where
+  foldr fn seed opt =
+    case opt of
+      Something a -> fn a seed
+      None        -> seed
+
+  foldMap fn opt =
+    case opt of
+      None        -> mempty
+      Something a -> fn a
+
+newtype Listy a = Listy [a]
```

Comment:
I think this newtype pattern is probably fine. We're going to need to use it for other types as well to avoid rebuilding them. We may just want to have a rep about newtypes and providing instances to lead up to this maybe.

Reply:
:+1: